pyHalo/realization_extensions.py

```diff
@@ -328,4 +377,102 @@ def add_primordial_black_holes(self, pbh_mass_fraction, kwargs_pbh_mass_function
+    :param shape: keyword argument for fluctuation geometry, see 'add_ULDM_fluctuations'
+    :param n_flucs: Number of fluctuations to make
+    :param args: properties of the given shape, see 'add_ULDM_fluctuations'
+    """
+
+    D_d = realization.lens_cosmo.cosmo.D_A_z(realization._zlens)
+    arcsec = realization.lens_cosmo.cosmo.arcsec
+    r_dB_angle = de_Broglie_wavelength / D_d / arcsec / 1e3 #de Broglie wavelength in arcsec
+
+    mass = RealizationDefaults().host_m200
+    c = realization.lens_cosmo.NFW_concentration(mass, realization._zlens)
+    _,_,r200 = realization.lens_cosmo.NFW_params_physical(mass,c,realization._zlens)
+    CLT_factor = np.sqrt(2*de_Broglie_wavelength/3/r200)
+
+    if shape!='aperture':
+
+        sigs = np.abs(np.random.normal(r_dB_angle,r_dB_angle/2,n_flucs)) #random widths
+        amps = CLT_factor*np.random.normal(0,np.sqrt(2)*rho_bar,n_flucs)*(2*np.pi*sigs**2) #random amplitudes
+
+    if shape=='ring':
+
+        angles = np.random.uniform(0,2*np.pi,n_flucs)  # random angles
+        radii = args['rmin'] + np.sqrt(np.random.uniform(0,1,n_flucs))*(args['rmax']-args['rmin']) #random radii
+        xs = radii*np.cos(angles) #random x positions
+        ys = radii*np.sin(angles) #random y positions
+
+    if shape=='ellipse':
+
+        angles = np.random.uniform(0,2*np.pi,n_flucs)  # random angles
+        aa = np.sqrt(np.random.uniform(0,1,n_flucs))*(args['amax']-args['amin']) + args['amin'] #random axis 1
+        bb = np.sqrt(np.random.uniform(0,1,n_flucs))*(args['bmax']-args['bmin']) + args['bmin'] #random axis 1
+        xs = aa*np.cos(angles)*np.cos(args['angle'])-bb*np.sin(angles)*np.sin(args['angle']) #random x positions
+        ys = aa*np.cos(angles)*np.sin(args['angle'])+bb*np.sin(angles)*np.cos(args['angle']) #random y positions
+
+    if shape=='aperture':
+        
+        amps,sigs,xs,ys=np.array([]),np.array([]),np.array([]),np.array([])
+
+        for i in range(len(n_flucs)): #loop through each image
+
+            sigs_i = np.abs(np.random.normal(r_dB_angle,r_dB_angle/2,n_flucs[i])) #random widths
+            amps_i = CLT_factor*np.random.normal(0,np.sqrt(2)*rho_bar,n_flucs[i])*(2*np.pi*sigs_i**2) #random amplitudes
+            angles_i = np.random.uniform(0,2*np.pi,n_flucs[i])  # random angles
+            radii_i = np.sqrt(np.random.uniform(0,1,n_flucs[i]))*args['aperture'] #random radii
+            xs_i = radii_i*np.cos(angles_i)+args['x_images'][i] #random x positions
+            ys_i = radii_i*np.sin(angles_i)+args['y_images'][i] #random y positions
+
+            amps,sigs,xs,ys=np.append(amps,amps_i),np.append(sigs,sigs_i),np.append(xs,xs_i),np.append(ys,ys_i)
 
+    args_fluc=[{'amp': amps[i], 'sigma': sigs[i], 'center_x': xs[i], 'center_y': ys[i]} for i in range(len(amps))]
+    fluctuations = [Gaussian(5*sigs[i], xs[i], ys[i], None, None, realization._zlens, None, realization.lens_cosmo,args_fluc[i],np.random.rand()) for i in range(len(amps))]
```

Comment:
can you relate the "mass" parameter to sigs and amps, and then internal to the Gaussian class convert back to sigs and amps? this would keep the concept of a halo mass common to each halo model (even if it's technically not an individual halo)

Reply:
That was a bug, it should have been:

masses=[GaussianKappa().mass_3d_lens(5*arg_fluc['sigma'],arg_fluc['amp'],arg_fluc['sigma']) for arg_fluc in args_fluc]

and then I use those masses instead of 5*sigs[i]. Do you want me to implement this mass definition in the Gaussian class?